src/main.rs

```diff
@@ -479,44 +534,65 @@ fn run() -> Result<(), failure::Error> {
         commands::subdomain(name, &user, &target)?;
     } else if let Some(kv_matches) = matches.subcommand_matches("kv:namespace") {
         let manifest = settings::target::Manifest::new(config_path)?;
-        let target = manifest.get_target(matches.value_of("env"), false)?;
-
         let user = settings::global_user::GlobalUser::new()?;
 
         match kv_matches.subcommand() {
             ("create", Some(create_matches)) => {
+                let target = manifest.get_target(create_matches.value_of("env"), false)?;
+
                 let title = create_matches.value_of("title").unwrap();
                 commands::kv::namespace::create(&target, user, title)?;
             }
             ("delete", Some(delete_matches)) => {
-                let id = delete_matches.value_of("namespace-id").unwrap();
-                commands::kv::namespace::delete(&target, user, id)?;
+                let target = manifest.get_target(delete_matches.value_of("env"), false)?;
+
+                let namespace_binding = delete_matches.value_of("namespace").unwrap();
+                let namespace_id = commands::kv::get_namespace_id(&target, namespace_binding);
+                match namespace_id {
+                    Some(namespace_id) => commands::kv::namespace::delete(&target, user, &namespace_id)?,
+                    None => failure::bail!("Namespace \"{}\" not found", namespace_binding),
+                }
             }
             ("rename", Some(rename_matches)) => {
-                let id = rename_matches.value_of("namespace-id").unwrap();
+                let target = manifest.get_target(rename_matches.value_of("env"), false)?;
+
+                let namespace_binding = rename_matches.value_of("namespace").unwrap();
+                let namespace_id = commands::kv::get_namespace_id(&target, namespace_binding);
                 let title = rename_matches.value_of("title").unwrap();
-                commands::kv::namespace::rename(&target, user, id, title)?;
+                match namespace_id {
+                    Some(namespace_id) => commands::kv::namespace::rename(&target, user, &namespace_id, title)?,
+                    None => failure::bail!("Namespace \"{}\" not found", namespace_binding),
+                }
             }
-            ("list", Some(_list_matches)) => {
+            ("list", Some(list_matches)) => {
+                let target = manifest.get_target(list_matches.value_of("env"), false)?;
+
                 commands::kv::namespace::list(&target, user)?;
             }
             ("", None) => message::warn("kv:namespace expects a subcommand"),
             _ => unreachable!(),
         }
     } else if let Some(kv_matches) = matches.subcommand_matches("kv:key") {
         let manifest = settings::target::Manifest::new(config_path)?;
-        let target = manifest.get_target(matches.value_of("env"), false)?;
```

Comment:
why pull this out of top level? it's still the same everywhere.

Reply:
We can't pull it out of top level because the argument corresponds to "env", which is an arg for `put`, `delete`, etc. (not `kv:key`)

Reply:
ah. got it.